[c++] Distance-agnostic triplet margin loss

ethch18 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45377 [c++] Distance-agnostic triplet margin loss**

This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee

Differential Revision: [D24003973](https://our.internmc.facebook.com/intern/diff/D24003973)

ethch18 committed:
[c++] Distance-agnostic triplet margin loss

[ghstack-poisoned]

ethch18 committed:
Update on "[c++] Distance-agnostic triplet margin loss"


This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee 

[ghstack-poisoned]

glaringlee commented:
@ethch18 
I'll rebase this soon. Can you do a rebase? It seems the CI tests failed due to a problematic master base.

ethch18 committed:
Update on "[c++] Distance-agnostic triplet margin loss"


This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee 

[ghstack-poisoned]

glaringlee commented on torch/csrc/api/src/nn/modules/loss.cpp:
```diff
@@ -180,6 +180,33 @@ Tensor TripletMarginLossImpl::forward(
 // ============================================================================
 
+TripletMarginWithDistanceLossImpl::TripletMarginWithDistanceLossImpl(
+    const TripletMarginWithDistanceLossOptions& options_) // NOLINT(modernize-pass-by-value)
```
Will complier complain if removing `NOLINT(...)`?
I kinda remember c10:optional doesn't work well with std::move. Did you try std::move here?

ethch18 replied:
clint complains about this otherwise; I copied the `NOLINT()` from elsewhere in the file.

What would `std::move` look like in this case?  Looks like everything else in this file uses the same format, minus `NOLINT()` if it's not applicable.

glaringlee replied:
You can do the following to avoid that lint error.
`
TripletMarginWithDistanceLossImpl::TripletMarginWithDistanceLossImpl(
    TripletMarginWithDistanceLossOptions options_) : 
    options(std::move(options_)) {}
`
But I kinda remember std::move is not working well with c10::optional which is inside options. But if you can try, that would be great.

ethch18 replied:
std::move couldn't build once for me locally, then it started building -- not sure if it's a flaky failure.  I've pushed this and if the CI fails, then I'll go back to pass-by-reference

ethch18 commented on test/cpp/api/modules.cpp:
```diff
@@ -2085,6 +2085,109 @@ TEST_F(ModulesTest, TripletMarginLoss) {
+        TripletMarginWithDistanceLoss distanceLoss(distanceOptions);
+
+        auto basicOutput = basicLoss->forward(anchor, positive, negative);
+        auto distanceOutput = distanceLoss->forward(anchor, positive, negative);
```
Does `()` mean, e.g., `distanceLoss(anchor, positive, negative)`?

glaringlee replied:
Ah, sry, I am not clear. Yes.  distanceLoss(anchor, positive, negative).

ethch18 committed:
Update on "[c++] Distance-agnostic triplet margin loss"


This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee 

[ghstack-poisoned]

glaringlee commented:
@ethch18 
All CI builds passed (rocm test failure is fine). I am going to build this on my end locally, will approve this if it passed. Thanks a lot for contributing.

ethch18 commented:
@glaringlee great!  FYI, Mike asked me to incorporate the changes from PR #45378 into this one too, which I'll do shortly

glaringlee commented:
@ethch18 sure, go ahead!

ethch18 committed:
Update on "[c++] Distance-agnostic triplet margin loss"


This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee 

[ghstack-poisoned]

ethch18 commented:
@glaringlee updated -- let me know if you need anything else from my end!

glaringlee commented:
@ethch18 
Thanks, can you please rebase your code once? There are several CI test failed, I think that's upstream problems, rebase should solve it.

ethch18 committed:
Update on "[c++] Distance-agnostic triplet margin loss"


This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee 

[ghstack-poisoned]

ethch18 commented:
Just rebased on fbcode/warm

glaringlee commented:
@ethch18 
Sry, can you rebase again, this code still failed one of our CI test: pytorch_xla_linux_bionic_py3_6_clang9_build.

ethch18 committed:
Update on "[c++] Distance-agnostic triplet margin loss"


This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee 

[ghstack-poisoned]

ethch18 commented:
@glaringlee rebased with fbcode/warm again.  If this doesn't work, I'll try viable/strict next

glaringlee commented:
@ethch18 
Thanks, it should work. There is a know issue for our XLA build, it can be solved by rebase.

glaringlee reviewed (approved):
LGTM now
